File: ios_dashboard/SandboxedDashboard/Views/Control/AskSheet.swift
import SwiftUI

/// Ask — the iOS surface for the non-interrupting sidecar co-pilot.
///
/// Presented as a bottom sheet (medium/large detents) over the mission. It runs
/// in its own lane: it never touches the mission's queue or the working agent.
/// Threads/messages live in a separate backend store and are rendered here with
/// a distinct cyan "co-pilot" identity.
struct AskSheet: View {
    let missionId: String
    /// Drop an Ask answer into the real mission composer (optional bridge).
    var onSendToAgent: ((String) -> Void)? = nil
    let onDismiss: () -> Void

    @State private var threads: [AskThread] = []
    @State private var threadId: String?
    @State private var messages: [AskMessage] = []
    @State private var input: String = ""
    @State private var isLoading = false
    @State private var errorText: String?
    // Id of the assistant bubble currently being streamed into (nil between segments).
    @State private var streamId: String?
    // Bumped on send / thread switch so a stale post-stream sync can be skipped.
    @State private var streamGen = 0

    private let api = APIService.shared
    private let copilot = Color.cyan

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                conversation
                composer
            }
            .background(Theme.backgroundSecondary)
            .navigationTitle("Ask")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar { toolbarContent }
            // Re-run when the mission changes while the sheet stays open: a
            // bare `.task` would keep the old mission's threads (and let a
            // superseded stream mutate the view). `id:` restarts it per mission.
            .task(id: missionId) {
                streamGen += 1
                isLoading = false
                streamId = nil
                threadId = nil
                messages = []
                threads = []
                errorText = nil
                await loadThreads()
            }
        }
    }

    // MARK: - Conversation

    private var conversation: some View {
        ScrollViewReader { proxy in
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 10) {
                    if messages.isEmpty && !isLoading {
                        emptyState
                    }
                    ForEach(messages) { message in
                        AskBubble(
                            message: message,
                            copilot: copilot,
                            onSendToAgent: onSendToAgent,
                            onRetry: message.sendState.isFailed
                                ? { Task { await retry(message) } } : nil
                        )
                        .id(message.id)
                    }
                    if isLoading {
                        HStack(spacing: 6) {
                            ProgressView().controlSize(.small)
                            Text("thinking…")
                                .font(.caption)
                                .foregroundStyle(copilot.opacity(0.8))
                        }
                        .id("loading")
                    }
                    if let errorText {
                        Text(errorText)
                            .font(.caption)
                            .foregroundStyle(Theme.error)
                            .padding(8)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Theme.error.opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                }
                .padding(16)
            }
            .onChange(of: messages.count) { _, _ in
                if let last = messages.last {
                    withAnimation { proxy.scrollTo(last.id, anchor: .bottom) }
                }
            }
            .onChange(of: isLoading) { _, loading in
                if loading { withAnimation { proxy.scrollTo("loading", anchor: .bottom) } }
            }
        }
    }

    private var emptyState: some View {
        VStack(spacing: 8) {
            Image(systemName: "sparkles")
                .font(.system(size: 22))
                .foregroundStyle(copilot.opacity(0.5))
            Text("Ask about this mission — what it's doing, why, or inspect the workspace. The working agent is never interrupted.")
                .font(.footnote)
                .foregroundStyle(Theme.textMuted)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(.top, 40)
    }

    // MARK: - Composer

    private var composer: some View {
        HStack(spacing: 8) {
            TextField("Ask the co-pilot…", text: $input, axis: .vertical)
                .lineLimit(1...4)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(Theme.card)
                .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
                .onSubmit { Task { await send() } }

            Button {
                Task { await send() }
            } label: {
                Image(systemName: "arrow.up.circle.fill")
                    .font(.system(size: 28))
                    .foregroundStyle(input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || isLoading ? Theme.textMuted : copilot)
            }
            .disabled(input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || isLoading)
        }
        .padding(12)
        .background(.ultraThinMaterial)
    }

    // MARK: - Toolbar

    @ToolbarContentBuilder
    private var toolbarContent: some ToolbarContent {
        ToolbarItem(placement: .topBarLeading) {
            Button("Done") { onDismiss() }
                .foregroundStyle(copilot)
        }
        ToolbarItem(placement: .topBarTrailing) {
            HStack(spacing: 14) {
                Menu {
                    Button {
                        newThread()
                    } label: {
                        Label("New thread", systemImage: "plus")
                    }
                    if !threads.isEmpty {
                        Divider()
                        ForEach(threads) { thread in
                            Button {
                                Task { await selectThread(thread.id) }
                            } label: {
                                Label(thread.displayTitle, systemImage: thread.id == threadId ? "checkmark" : "bubble.left")
                            }
                        }
                    }
                } label: {
                    Image(systemName: "bubble.left.and.bubble.right")
                }

                Button(role: .destructive) {
                    Task { await clearThread() }
                } label: {
                    Image(systemName: "trash")
                }
                .disabled(threadId == nil)
            }
            .foregroundStyle(copilot)
        }
    }

    // MARK: - Actions

    private func loadThreads() async {
        do {
            let fetched = try await api.listAskThreads(missionId: missionId)
            threads = fetched
            if let first = fetched.first {
                await selectThread(first.id)
            }
        } catch {
            // Non-fatal — just start with an empty thread.
        }
    }

    private func selectThread(_ id: String) async {
        streamGen += 1
        let gen = streamGen
        isLoading = false
        streamId = nil
        threadId = id
        do {
            let detail = try await api.getAskThread(missionId: missionId, threadId: id)
            // A later switch / send may have superseded this fetch.
            if gen == streamGen { messages = detail.messages }
        } catch {
            if gen == streamGen { messages = [] }
        }
    }

    private func newThread() {
        streamGen += 1
        isLoading = false
        streamId = nil
        threadId = nil
        messages = []
        input = ""
        errorText = nil
    }

    private func send() async {
        let content = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty, !isLoading else { return }
        input = ""
        let userId = "u-\(UUID().uuidString)"
        messages.append(
            AskMessage(
                id: userId,
                threadId: threadId ?? "",
                seq: messages.count + 1,
                role: "user",
                content: content,
                toolName: nil,
                toolCallId: nil,
                createdAt: isoNow(),
                sendState: .pending
            )
        )
        await runTurn(userMessageId: userId, content: content)
    }

    /// Re-run a co-pilot turn for a user message whose send failed. Reuses the
    /// existing bubble (no duplicate) with the same rollback semantics.
    private func retry(_ message: AskMessage) async {
        guard message.isUser, message.sendState.isFailed, !isLoading else { return }
        await runTurn(userMessageId: message.id, content: message.content)
    }

    /// Drives one streamed co-pilot turn. The user bubble is **preserved**
    /// across failures (flipped to `.failed` with a tap-to-retry) instead of
    /// being rolled back — mirroring the main mission composer so a dropped
    /// message never silently vanishes. Only this turn's streamed
    /// assistant/tool bubbles roll back on failure.
    private func runTurn(userMessageId: String, content: String) async {
        errorText = nil
        isLoading = true
        streamId = nil
        streamGen += 1
        let gen = streamGen
        if let i = messages.firstIndex(where: { $0.id == userMessageId }) {
            messages[i].sendState = .pending
        }
        // Roll-back boundary: keep everything up to and including the user
        // bubble; drop streamed bubbles appended during this turn on failure.
        let baseCount =
            messages.firstIndex(where: { $0.id == userMessageId }).map { $0 + 1 }
            ?? messages.count

        var failure: String?
        do {
            for try await ev in api.askStream(
                missionId: missionId,
                content: content,
                threadId: threadId
            ) {
                // A newer send / thread switch superseded this turn.
                if gen != streamGen { return }
                // First event back means the backend accepted the message.
                if let i = messages.firstIndex(where: { $0.id == userMessageId }),
                    messages[i].sendState.isPending
                {
                    messages[i].sendState = .sent
                }
                handleStreamEvent(ev)
            }
        } catch {
            if gen == streamGen { failure = error.localizedDescription }
        }

        // Superseded — a newer turn owns the list and the loading flag now.
        if gen != streamGen { return }

        if let failure {
            // Preserve the user bubble (failed + retry); drop streamed bubbles.
            if messages.count > baseCount {
                messages = Array(messages.prefix(baseCount))
            }
            if let i = messages.firstIndex(where: { $0.id == userMessageId }) {
                messages[i].sendState = .failed(reason: failure)
            }
        } else if let i = messages.firstIndex(where: { $0.id == userMessageId }) {
            messages[i].sendState = .sent
        }
        isLoading = false
        streamId = nil
    }

    private func isoNow() -> String {
        ISO8601DateFormatter().string(from: Date())
    }

    private func handleStreamEvent(_ ev: AskStreamEvent) {
        switch ev.type {
        case "delta":
            guard let text = ev.content else { return }
            if let id = streamId,
                let idx = messages.firstIndex(where: { $0.id == id })
            {
                let m = messages[idx]
                messages[idx] = AskMessage(
                    id: m.id,
                    threadId: m.threadId,
                    seq: m.seq,
                    role: m.role,
                    content: m.content + text,
                    toolName: nil,
                    toolCallId: nil,
                    createdAt: m.createdAt
                )
            } else {
                let id = "a-\(UUID().uuidString)"
                streamId = id
                messages.append(
                    AskMessage(
                        id: id,
                        threadId: threadId ?? "",
                        seq: messages.count + 1,
                        role: "assistant",
                        content: text,
                        toolName: nil,
                        toolCallId: nil,
                        createdAt: isoNow()
                    )
                )
            }
        case "tool_call":
            streamId = nil
            messages.append(
                AskMessage(
                    id: "tc-\(ev.toolCallId ?? UUID().uuidString)",
                    threadId: threadId ?? "",
                    seq: messages.count + 1,
                    role: "tool_call",
                    content: ev.args ?? "",
                    toolName: ev.name,
                    toolCallId: ev.toolCallId,
                    createdAt: isoNow()
                )
            )
        case "tool_result":
            messages.append(
                AskMessage(
                    id: "tr-\(ev.toolCallId ?? UUID().uuidString)",
                    threadId: threadId ?? "",
                    seq: messages.count + 1,
                    role: "tool_result",
                    content: ev.result ?? "",
                    toolName: ev.name,
                    toolCallId: ev.toolCallId,
                    createdAt: isoNow()
                )
            )
        case "done":
            streamId = nil
            let tid = ev.threadId
            if let tid { threadId = tid }
            let gen = streamGen
            Task {
                if let refreshed = try? await api.listAskThreads(missionId: missionId),
                    gen == streamGen
                {
                    threads = refreshed
                }
                // Reconcile the streamed bubbles with the canonical persisted
                // messages, unless a newer send / thread switch superseded this.
                if let tid,
                    let detail = try? await api.getAskThread(missionId: missionId, threadId: tid),
                    gen == streamGen, threadId == tid
                {
                    messages = detail.messages
                }
            }
        // "error" events are surfaced via the throw path in askStream, so they
        // flow through send()'s catch (gen-guarded rollback + restore).
        default:
            break
        }
    }

    private func clearThread() async {
        guard let id = threadId else { return }
        try? await api.deleteAskThread(missionId: missionId, threadId: id)
        if let refreshed = try? await api.listAskThreads(missionId: missionId) {
            threads = refreshed
        }
        newThread()
    }
}

// MARK: - Bubble

private struct AskBubble: View {
    let message: AskMessage
    let copilot: Color
    var onSendToAgent: ((String) -> Void)?
    var onRetry: (() -> Void)? = nil

    var body: some View {
        if message.isUser {
            HStack {
                Spacer(minLength: 40)
                VStack(alignment: .trailing, spacing: 4) {
                    Text(message.content)
                        .font(.subheadline)
                        .foregroundStyle(Theme.textPrimary)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 8)
                        .background(
                            message.sendState.isFailed
                                ? Theme.error.opacity(0.18) : Theme.card
                        )
                        .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
                        .overlay(
                            RoundedRectangle(cornerRadius: 14, style: .continuous)
                                .stroke(
                                    message.sendState.isFailed ? Theme.error : Color.clear,
                                    lineWidth: 1
                                )
                        )
                        // Dim + spinner while awaiting backend ack; this is the
                        // co-pilot peer of the main composer's pending bubble.
                        .opacity(message.sendState.isPending ? 0.55 : 1)
                        .overlay(alignment: .bottomTrailing) {
                            if message.sendState.isPending {
                                ProgressView()
                                    .controlSize(.mini)
                                    .padding(6)
                            } else if message.sendState.isFailed {
                                Image(systemName: "exclamationmark.circle.fill")
                                    .font(.caption)
                                    .foregroundStyle(Theme.error)
                                    .padding(6)
                            }
                        }
                        .animation(.easeOut(duration: 0.15), value: message.sendState.isPending)
                        .animation(.easeOut(duration: 0.15), value: message.sendState.isFailed)
                    // Inline "Not sent · Tap to retry" — preserves the message
                    // on screen instead of dropping it on a failed turn.
                    if message.sendState.isFailed {
                        Button { onRetry?() } label: {
                            HStack(spacing: 4) {
                                Image(systemName: "arrow.clockwise.circle.fill")
                                    .font(.caption2)
                                Text("Not sent · Tap to retry")
                                    .font(.caption2.weight(.medium))
                            }
                            .foregroundStyle(Theme.error)
                        }
                        .buttonStyle(.plain)
                    }
                }
            }
        } else if message.isTool {
            HStack(alignment: .top, spacing: 6) {
                Image(systemName: "terminal")
                    .font(.system(size: 10))
                    .foregroundStyle(Theme.textMuted)
                Text(toolSummary)
                    .font(.system(size: 11, design: .monospaced))
                    .foregroundStyle(Theme.textMuted)
                    .lineLimit(3)
            }
            .padding(.leading, 24)
            .frame(maxWidth: .infinity, alignment: .leading)
        } else {
            // assistant
            HStack(alignment: .top, spacing: 8) {
                Image(systemName: "sparkles")
                    .font(.system(size: 13))
                    .foregroundStyle(copilot)
                    .padding(.top, 2)
                VStack(alignment: .leading, spacing: 6) {
                    Text(message.content)
                        .font(.subheadline)
                        .foregroundStyle(Theme.textPrimary)
                        .textSelection(.enabled)
                    if let onSendToAgent {
                        Button {
                            onSendToAgent(message.content)
                        } label: {
                            Label("Send to agent", systemImage: "arrow.uturn.left")
                                .font(.caption2)
                                .foregroundStyle(copilot.opacity(0.8))
                        }
                    }
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(copilot.opacity(0.08))
                .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
            }
        }
    }

    private var toolSummary: String {
        let label = message.toolName.map { "\($0) → " } ?? "↳ "
        let body: String
        if message.isToolCall, let data = message.content.data(using: .utf8),
           let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            body = (obj["command"] as? String) ?? (obj["path"] as? String) ?? message.content
        } else {
            body = message.content
        }
        let trimmed = body.count > 200 ? String(body.prefix(200)) + "…" : body
        return label + trimmed
    }
}

File: android_dashboard/app/src/main/java/sh/sandboxed/dashboard/ui/control/AskSheet.kt
package sh.sandboxed.dashboard.ui.control

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Reply
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.ErrorOutline
import androidx.compose.material.icons.filled.Forum
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Terminal
import androidx.compose.material3.BottomSheetDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import org.json.JSONObject
import sh.sandboxed.dashboard.data.AppContainer
import sh.sandboxed.dashboard.data.AskMessage
import sh.sandboxed.dashboard.data.AskThread
import sh.sandboxed.dashboard.ui.TestTags
import sh.sandboxed.dashboard.ui.tag
import sh.sandboxed.dashboard.ui.theme.Palette
import sh.sandboxed.dashboard.util.Haptics

// Cyan "co-pilot" identity, distinct from the indigo accent of the main agent.
private val Copilot = Color(0xFF22D3EE)

/// Ask — the Android surface for the non-interrupting sidecar co-pilot.
///
/// A bottom sheet over the mission that runs in its own lane: it never touches
/// the mission's queue or the working agent. Threads/messages live in a
/// separate backend store and render here with a distinct cyan identity.
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AskSheet(
    container: AppContainer,
    missionId: String,
    onSendToAgent: (String) -> Unit,
    onDismiss: () -> Unit,
) {
    val vm = remember(missionId) { AskViewModel(container, missionId) }
    val state by vm.state.collectAsState()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val haptics = remember { Haptics(container) }
    val listState = rememberLazyListState()
    var input by remember { mutableStateOf("") }

    DisposableEffect(vm) { onDispose { vm.stop() } }

    LaunchedEffect(state.messages.size, state.isLoading) {
        val count = state.messages.size + if (state.isLoading) 1 else 0
        if (count > 0) listState.animateScrollToItem(count - 1)
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = Palette.BackgroundSecondary,
        dragHandle = { BottomSheetDefaults.DragHandle() },
    ) {
        Column(Modifier.fillMaxWidth().fillMaxHeight(0.92f).imePadding()) {
            AskHeader(
                threads = state.threads,
                threadId = state.threadId,
                onNewThread = { haptics.selection(); vm.newThread() },
                onSelectThread = { haptics.selection(); vm.selectThread(it) },
                onClear = { haptics.error(); vm.clearThread() },
            )
            LazyColumn(
                state = listState,
                modifier = Modifier.weight(1f).fillMaxWidth(),
                contentPadding = PaddingValues(16.dp),
                verticalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                if (state.messages.isEmpty() && !state.isLoading) {
                    item("empty") { AskEmptyState() }
                }
                items(state.messages, key = { it.id }) { message ->
                    AskBubble(
                        message = message,
                        onSendToAgent = { haptics.selection(); onSendToAgent(it) },
                        onRetry = if (message.sendState.isFailed) ({ vm.retry(message) }) else null,
                    )
                }
                if (state.isLoading) {
                    item("loading") { ThinkingRow() }
                }
                state.error?.let { err ->
                    item("error") {
                        Text(
                            err,
                            color = Palette.Error,
                            style = MaterialTheme.typography.bodySmall,
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(Palette.Error.copy(alpha = 0.1f), RoundedCornerShape(8.dp))
                                .padding(8.dp),
                        )
                    }
                }
            }
            AskComposer(
                value = input,
                onChange = { input = it },
                enabled = !state.isLoading,
                onSend = {
                    val text = input
                    input = ""
                    haptics.medium()
                    vm.send(text)
                },
            )
        }
    }
}

@Composable
private fun AskHeader(
    threads: List<AskThread>,
    threadId: String?,
    onNewThread: () -> Unit,
    onSelectThread: (String) -> Unit,
    onClear: () -> Unit,
) {
    var menuOpen by remember { mutableStateOf(false) }
    Row(
        Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(Icons.Filled.AutoAwesome, null, tint = Copilot, modifier = Modifier.size(18.dp))
        Spacer(Modifier.width(8.dp))
        Text("Ask", style = MaterialTheme.typography.titleMedium, color = Palette.TextPrimary)
        Spacer(Modifier.weight(1f))
        Box {
            IconButton(onClick = { menuOpen = true }, modifier = Modifier.tag(TestTags.ASK_THREADS)) {
                Icon(Icons.Filled.Forum, "Threads", tint = Copilot)
            }
            DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                DropdownMenuItem(
                    text = { Text("New thread") },
                    onClick = { menuOpen = false; onNewThread() },
                    leadingIcon = { Icon(Icons.Filled.Add, null) },
                    modifier = Modifier.tag(TestTags.ASK_NEW_THREAD),
                )
                threads.forEach { thread ->
                    DropdownMenuItem(
                        text = { Text(thread.displayTitle, maxLines = 1, overflow = TextOverflow.Ellipsis) },
                        onClick = { menuOpen = false; onSelectThread(thread.id) },
                        leadingIcon = {
                            if (thread.id == threadId) Icon(Icons.Filled.Check, null, tint = Copilot)
                            else Icon(Icons.Filled.Forum, null)
                        },
                    )
                }
            }
        }
        IconButton(
            onClick = onClear,
            enabled = threadId != null,
            modifier = Modifier.tag(TestTags.ASK_CLEAR),
        ) {
            Icon(Icons.Filled.Delete, "Clear thread", tint = if (threadId != null) Copilot else Palette.TextMuted)
        }
    }
}

@Composable
private fun AskEmptyState() {
    Column(
        Modifier.fillMaxWidth().padding(top = 40.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Icon(Icons.Filled.AutoAwesome, null, tint = Copilot.copy(alpha = 0.5f), modifier = Modifier.size(22.dp))
        Text(
            "Ask about this mission — what it's doing, why, or inspect the workspace. The working agent is never interrupted.",
            style = MaterialTheme.typography.bodySmall,
            color = Palette.TextMuted,
            modifier = Modifier.padding(horizontal = 24.dp),
        )
    }
}

@Composable
private fun ThinkingRow() {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        CircularProgressIndicator(modifier = Modifier.size(14.dp), strokeWidth = 2.dp, color = Copilot)
        Text("thinking…", style = MaterialTheme.typography.bodySmall, color = Copilot.copy(alpha = 0.8f))
    }
}

@Composable
private fun AskBubble(
    message: AskMessage,
    onSendToAgent: (String) -> Unit,
    onRetry: (() -> Unit)?,
) {
    when {
        message.isUser -> {
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
                Column(horizontalAlignment = Alignment.End, modifier = Modifier.widthIn(max = 320.dp)) {
                    Box(
                        Modifier
                            .background(
                                if (message.sendState.isFailed) Palette.Error.copy(alpha = 0.18f) else Palette.Card,
                                RoundedCornerShape(14.dp),
                            )
                            .padding(horizontal = 12.dp, vertical = 8.dp),
                    ) {
                        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                            Text(
                                message.content,
                                style = MaterialTheme.typography.bodyMedium,
                                color = if (message.sendState.isPending) Palette.TextMuted else Palette.TextPrimary,
                            )
                            if (message.sendState.isPending) {
                                CircularProgressIndicator(modifier = Modifier.size(12.dp), strokeWidth = 2.dp, color = Palette.TextMuted)
                            } else if (message.sendState.isFailed) {
                                Icon(Icons.Filled.ErrorOutline, "Failed", tint = Palette.Error, modifier = Modifier.size(14.dp))
                            }
                        }
                    }
                    if (message.sendState.isFailed && onRetry != null) {
                        Row(
                            Modifier.tag(TestTags.ASK_RETRY).clickable(onClick = onRetry).padding(top = 2.dp),
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(4.dp),
                        ) {
                            Icon(Icons.Filled.Refresh, null, tint = Palette.Error, modifier = Modifier.size(13.dp))
                            Text("Not sent · Tap to retry", style = MaterialTheme.typography.labelSmall, color = Palette.Error)
                        }
                    }
                }
            }
        }
        message.isTool -> {
            Row(
                Modifier.fillMaxWidth().padding(start = 24.dp),
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                Icon(Icons.Filled.Terminal, null, tint = Palette.TextMuted, modifier = Modifier.size(12.dp).padding(top = 2.dp))
                Text(
                    toolSummary(message),
                    style = MaterialTheme.typography.bodySmall,
                    fontFamily = FontFamily.Monospace,
                    color = Palette.TextMuted,
                    maxLines = 3,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
        else -> {
            // assistant
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                Icon(Icons.Filled.AutoAwesome, null, tint = Copilot, modifier = Modifier.size(14.dp).padding(top = 2.dp))
                Column(
                    Modifier
                        .widthIn(max = 320.dp)
                        .background(Copilot.copy(alpha = 0.08f), RoundedCornerShape(14.dp))
                        .padding(horizontal = 12.dp, vertical = 8.dp),
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(message.content, style = MaterialTheme.typography.bodyMedium, color = Palette.TextPrimary)
                    if (message.content.isNotBlank()) {
                        Row(
                            Modifier.tag(TestTags.ASK_SEND_TO_AGENT).clickable { onSendToAgent(message.content) },
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(4.dp),
                        ) {
                            Icon(Icons.AutoMirrored.Filled.Reply, null, tint = Copilot.copy(alpha = 0.8f), modifier = Modifier.size(13.dp))
                            Text("Send to agent", style = MaterialTheme.typography.labelSmall, color = Copilot.copy(alpha = 0.8f))
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun AskComposer(
    value: String,
    onChange: (String) -> Unit,
    enabled: Boolean,
    onSend: () -> Unit,
) {
    val canSend = value.isNotBlank() && enabled
    Row(
        Modifier.fillMaxWidth().background(Palette.BackgroundTertiary).padding(12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        OutlinedTextField(
            value = value,
            onValueChange = onChange,
            placeholder = { Text("Ask the co-pilot…", color = Palette.TextMuted) },
            maxLines = 4,
            modifier = Modifier.weight(1f).tag(TestTags.ASK_INPUT),
            colors = TextFieldDefaults.colors(
                focusedContainerColor = Palette.Card,
                unfocusedContainerColor = Palette.Card,
                focusedTextColor = Palette.TextPrimary,
                unfocusedTextColor = Palette.TextPrimary,
                cursorColor = Copilot,
            ),
        )
        IconButton(
            onClick = onSend,
            enabled = canSend,
            modifier = Modifier.tag(TestTags.ASK_SEND),
        ) {
            Icon(
                Icons.AutoMirrored.Filled.Send,
                "Send",
                tint = if (canSend) Copilot else Palette.TextMuted,
            )
        }
    }
}

private fun toolSummary(message: AskMessage): String {
    val label = message.toolName?.let { "$it → " } ?: "↳ "
    val body = if (message.isToolCall) {
        runCatching {
            val obj = JSONObject(message.content)
            obj.optString("command").takeIf { it.isNotBlank() }
                ?: obj.optString("path").takeIf { it.isNotBlank() }
                ?: message.content
        }.getOrDefault(message.content)
    } else {
        message.content
    }
    val trimmed = if (body.length > 200) body.take(200) + "…" else body
    return label + trimmed
}
